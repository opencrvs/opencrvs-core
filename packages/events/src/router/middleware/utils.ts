/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import type { AnyTRPCMiddlewareFunction } from '@trpc/server'
import {
  ActionInputWithType,
  TokenWithBearer,
  TokenUserType
} from '@opencrvs/commons'

export interface UserDetails {
  type: TokenUserType.USER | TokenUserType.SYSTEM
  id: string
  primaryOfficeId: string | undefined
  role: string
}

export type Context = { token: TokenWithBearer; user: UserDetails }
/**
 * TRPC Middleware options with correct context.
 * Actual middleware type definition is only for internal use within TRPC.
 */
export type MiddlewareOptions = Omit<
  Parameters<AnyTRPCMiddlewareFunction>[0],
  'ctx'
> & { ctx: Context }

export type ActionMiddlewareOptions = Omit<MiddlewareOptions, 'input'> & {
  input: ActionInputWithType
}
