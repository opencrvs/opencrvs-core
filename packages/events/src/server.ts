/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { createServer, IncomingMessage } from 'http'
import { createOpenApiHttpHandler } from 'trpc-to-openapi'
import { createHTTPHandler } from '@trpc/server/adapters/standalone'
import '@opencrvs/commons/monitoring'
import { logger } from '@opencrvs/commons'
import { appRouter } from './router/router'
import { createContext } from './context'

function stringifyRequest(req: IncomingMessage) {
  const url = new URL(req.url || '', `http://${req.headers.host}`)
  return `'${req.method} ${url.pathname}'`
}

const trpcConfig: Parameters<typeof createHTTPHandler>[0] = {
  router: appRouter,
  middleware: (req, _, next) => {
    logger.info(`Request: ${stringifyRequest(req)}`)
    return next()
  },
  onError: ({ error, req }) => {
    logger.warn(
      `Error for request: ${stringifyRequest(req)}. Error: '${error.message}'`
    )
  },
  createContext
}

function isTrpcRequest(req: IncomingMessage): boolean {
  if (!req.url) {
    throw new Error('No URL provided')
  }
  const url = new URL(req.url, `http://${req.headers.host}`)
  const pathName = url.pathname.replace(/^\//, '')
  const trpcProcedurePaths = Object.keys(appRouter._def.procedures)

  return (
    url.search.startsWith('?input') || trpcProcedurePaths.includes(pathName)
  )
}

export function server() {
  const restServer = createOpenApiHttpHandler(trpcConfig)
  const trpcServer = createHTTPHandler(trpcConfig)

  return createServer((req, res) => {
    if (!req.url) {
      res.writeHead(500)
      res.end('No URL provided')
      return
    }

    // If it's a tRPC request, handle it with the tRPC server
    if (isTrpcRequest(req)) {
      trpcServer(req, res)
    } else {
      // If it's a REST request, handle it with the REST server
      void restServer(req, res)
    }
  })
}
