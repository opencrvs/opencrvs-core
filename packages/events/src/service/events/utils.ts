/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { Action, ActionType } from '@opencrvs/commons'

export function findLastAssignmentAction(actions: Action[]) {
  return actions
    .filter(
      ({ type }) => type === ActionType.ASSIGN || type === ActionType.UNASSIGN
    )
    .reduce<
      Action | undefined
    >((latestAction, action) => (!latestAction || action.createdAt > latestAction.createdAt ? action : latestAction), undefined)
}
