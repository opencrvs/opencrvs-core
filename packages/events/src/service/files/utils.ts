/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { differenceWith, first, isEqual, orderBy } from 'lodash'
import {
  Draft,
  EventConfig,
  EventDocument,
  FieldConfig,
  FileFieldType,
  getActionAnnotationFields,
  AnyFileField,
  getDeclarationFields,
  getMixedPath,
  FileFieldWithOptionValue,
  FileFieldValue,
  ActionUpdate,
  getCurrentEventStateWithDrafts,
  EventState,
  getCurrentEventState,
  ActionDocument
} from '@opencrvs/commons'

function isFieldWithFile(field: FieldConfig): field is AnyFileField {
  return FileFieldType.safeParse(field.type).success
}

/**
 *
 * @param field File field
 * @param update single action or event state.
 * @returns normalised file reference based on the field id and event state.
 */
function getFileReferences(
  field: AnyFileField,
  update: ActionUpdate | EventState
): { id: string; value: string }[] {
  const value = getMixedPath<AnyFileField['defaultValue']>(update, field.id)

  if (!value) {
    return []
  }

  const maybeFileOptions = FileFieldWithOptionValue.safeParse(value)
  if (maybeFileOptions.success) {
    return maybeFileOptions.data.map((opt) => ({
      id: field.id,
      value: opt.path
    }))
  }

  const maybeFile = FileFieldValue.safeParse(value)

  if (maybeFile.success) {
    return [{ id: field.id, value: maybeFile.data.path }]
  }

  if (typeof value === 'string') {
    return [{ id: field.id, value }]
  }

  throw new Error(
    `Unknown filetype, ${field.type} with value: ${JSON.stringify(value)}`
  )
}

/**
 *  Since API allows sending PATCH requests with only the changed fields,
 *  we need to check the current event state with the current draft.
 * @returns file references that are not present in the current event state but are when previous draft is applied to the state.
 */
function getUnreferencedDeclarationDraftFiles({
  event,
  currentDraft,
  previousDraft,
  configuration
}: {
  event: EventDocument
  currentDraft?: Draft
  previousDraft: Draft
  configuration: EventConfig
}) {
  const currentEventState = currentDraft
    ? getCurrentEventStateWithDrafts({
        event,
        draft: currentDraft,
        configuration
      })
    : getCurrentEventState(event, configuration)

  const previousEventState = getCurrentEventStateWithDrafts({
    event,

    draft: {
      // Given previous draft, if event has been updated afterwards (running deletion after submission of actual action.) the data would never get applied.
      ...previousDraft,
      createdAt: new Date().toISOString(),
      action: {
        ...previousDraft.action,
        createdAt: new Date().toISOString()
      }
    },
    configuration
  })

  const declarationFileFields =
    getDeclarationFields(configuration).filter(isFieldWithFile)

  const currentDeclarationFiles = declarationFileFields.flatMap((field) =>
    getFileReferences(field, currentEventState.declaration)
  )

  const previousDeclarationFiles = declarationFileFields.flatMap((field) =>
    getFileReferences(field, previousEventState.declaration)
  )

  return differenceWith(
    previousDeclarationFiles,
    currentDeclarationFiles,
    isEqual
  )
}

function getUnreferencedAnnotationDraftFiles({
  currentAction,
  previousDraft,
  configuration
}: {
  currentAction: Draft['action'] | ActionDocument
  previousDraft: Draft
  configuration: EventConfig
}) {
  if (currentAction.type !== previousDraft.action.type) {
    // eslint-disable-next-line no-console
    console.warn(
      'Previous draft was for different action. Not comparing file references between them.'
    )
    return []
  }

  const currentDraftActionConfig = configuration.actions.find(
    (action) => action.type === previousDraft.action.type
  )

  if (!currentDraftActionConfig) {
    // eslint-disable-next-line no-console
    console.warn(
      `No configuration found for action ${previousDraft.action.type}.`
    )
  }

  const annotationFileFields = currentDraftActionConfig
    ? getActionAnnotationFields(currentDraftActionConfig).filter(
        isFieldWithFile
      )
    : []

  // @TODO: There might be a case where annotation shares id with declaration field? Should it limited by the configuration?
  const currentAnnotationFiles = annotationFileFields.flatMap((field) =>
    getFileReferences(field, currentAction.annotation ?? {})
  )

  const previousAnnotationFiles = annotationFileFields.flatMap((field) =>
    getFileReferences(field, previousDraft.action.annotation ?? {})
  )

  return differenceWith(
    previousAnnotationFiles,
    currentAnnotationFiles,
    isEqual
  )
}

/**
 *
 * @returns all the filenames that are referenced in the current draft, but not in the previous draft.
 */
export function getUnreferencedDraftFiles({
  event,
  currentDraft,
  previousDraft,
  configuration
}: {
  event: EventDocument
  /**
   * Current draft is optional, because we might be comparing the previous draft with the current event state. (e.g. actual action has been submitted)
   */
  currentDraft?: Draft
  previousDraft: Draft
  configuration: EventConfig
}) {
  const unreferencedDeclarationDraftFiles =
    getUnreferencedDeclarationDraftFiles({
      event,
      currentDraft,
      previousDraft,
      configuration
    })

  const currentAction =
    currentDraft?.action ??
    first(
      orderBy(
        event.actions.filter(
          (action) => action.type === previousDraft.action.type
        ),
        ['createdAt'],
        'desc'
      )
    )

  if (!currentAction) {
    // This should not happen. It means that the previous draft action type was invalid.
    // eslint-disable-next-line no-console
    console.error(
      `Could not find corresponding action ${previousDraft.action.type} for event: ${event.id}`
    )

    return unreferencedDeclarationDraftFiles
  }

  const unreferencedAnnotationDraftFiles = getUnreferencedAnnotationDraftFiles({
    // @TODO: for some reason declaration is omitted from some type?
    currentAction: currentAction as Draft['action'] | ActionDocument,
    previousDraft,
    configuration
  })

  return [
    ...unreferencedDeclarationDraftFiles,
    ...unreferencedAnnotationDraftFiles
  ]
}
