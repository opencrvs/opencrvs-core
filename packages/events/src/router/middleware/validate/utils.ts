/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { TRPCError } from '@trpc/server'
import _ from 'lodash'
import {
  ActionUpdate,
  FieldConfig,
  getFieldValidationErrors,
  Inferred,
  errorMessages
} from '@opencrvs/commons/events'

type ValidationError = {
  message: string
  id: string
  value: unknown
}

export function getFormFieldErrors(formFields: Inferred[], data: ActionUpdate) {
  return formFields.reduce(
    (errorResults: ValidationError[], field: FieldConfig) => {
      const fieldErrors = getFieldValidationErrors({
        field,
        values: data
      }).errors

      if (fieldErrors.length === 0) {
        return errorResults
      }

      // For backend, use the default message without translations.
      const errormessageWithId = fieldErrors.map((error) => ({
        message: error.message.defaultMessage,
        id: field.id,
        value: data[field.id as keyof typeof data]
      }))

      return [...errorResults, ...errormessageWithId]
    },
    []
  )
}

export function getVerificationPageErrors(
  verificationPageIds: string[],
  data: ActionUpdate
) {
  return verificationPageIds
    .map((pageId) => {
      const value = data[pageId]
      return typeof value !== 'boolean'
        ? {
            message: 'Verification page result is required',
            id: pageId,
            value
          }
        : null
    })
    .filter((error) => error !== null)
}

export function throwWhenNotEmpty(errors: unknown[]) {
  if (errors.length > 0) {
    throw new TRPCError({
      code: 'BAD_REQUEST',
      message: JSON.stringify(errors)
    })
  }
}

function isObject(value: unknown): value is Record<string, unknown> {
  return _.isPlainObject(value)
}

/**
 * Flattens an object into a list of [key, value] pairs using dot/bracket notation.
 */
function flattenEntries<T>(obj: T, path: string = ''): [string, unknown][] {
  if (Array.isArray(obj)) {
    return obj.flatMap((value, index) =>
      flattenEntries(value, `${path}[${index}]`)
    )
  }

  if (isObject(obj)) {
    return Object.entries(obj).flatMap(([key, value]) =>
      flattenEntries(value, path ? `${path}.${key}` : key)
    )
  }

  return obj === undefined ? [] : [[path, obj]]
}

/**
 * Compares update vs cleaned payloads and returns an array of offending keys.
 */
export function getInvalidUpdateKeys<T>({
  update,
  cleaned
}: {
  update: T
  cleaned: T
}): ValidationError[] {
  const updateEntries = flattenEntries(update)
  const cleanedKeys = flattenEntries(cleaned).map(([key]) => key)

  return updateEntries
    .filter(([key]) => !cleanedKeys.includes(key))
    .map(([key, value]) => ({
      message: errorMessages.hiddenField.defaultMessage,
      id: key,
      value
    }))
}
