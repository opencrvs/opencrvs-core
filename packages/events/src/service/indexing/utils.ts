/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { EventSearchIndex } from '@opencrvs/commons/events'
import {
  QueryDslQueryContainer,
  SearchRequest
} from '@elastic/elasticsearch/lib/api/types'

const FIELD_SEPARATOR = '____'
export const DEFAULT_SIZE = 10

/**
 * Generates an Elasticsearch query to search within `document.data`
 * using the provided search payload.
 */
export function generateQuery(event: Omit<EventSearchIndex, 'type'>) {
  const must: QueryDslQueryContainer[] = Object.entries(event).map(
    ([key, value]) => ({
      match: {
        [`data.${key.replaceAll('.', FIELD_SEPARATOR)}`]: value
      }
    })
  )

  return {
    bool: {
      must
    }
  } as SearchRequest['query']
}
