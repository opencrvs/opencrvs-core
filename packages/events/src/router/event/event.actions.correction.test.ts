/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { TRPCError } from '@trpc/server'
import {
  ActionType,
  AddressType,
  createPrng,
  EventDocument,
  generateActionDeclarationInput,
  getAcceptedActions,
  getUUID,
  SCOPES
} from '@opencrvs/commons'
import { tennisClubMembershipEvent } from '@opencrvs/commons/fixtures'
import { createTestClient, setupTestCase } from '@events/tests/utils'

test(`${ActionType.REQUEST_CORRECTION} prevents forbidden access if missing required scope`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user, [])

  await expect(
    client.event.actions.correction.request(
      generator.event.actions.correction.request(
        'registered-event-test-id-12345'
      )
    )
  ).rejects.toMatchObject(new TRPCError({ code: 'FORBIDDEN' }))
})

test(`${ActionType.REQUEST_CORRECTION} allows access if required scope is present`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user, [
    SCOPES.RECORD_REGISTRATION_REQUEST_CORRECTION
  ])

  await expect(
    client.event.actions.correction.request(
      generator.event.actions.correction.request(
        'registered-event-test-id-12345'
      )
    )
  ).rejects.not.toMatchObject(new TRPCError({ code: 'FORBIDDEN' }))
})

test(`${ActionType.APPROVE_CORRECTION} prevents forbidden access if missing required scope`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user, [])

  await expect(
    client.event.actions.correction.approve(
      generator.event.actions.correction.approve(
        'registered-event-test-id-12345',
        'request-test-id-12345'
      )
    )
  ).rejects.toMatchObject(new TRPCError({ code: 'FORBIDDEN' }))
})

test(`${ActionType.APPROVE_CORRECTION} allows access if required scope is present`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user, [SCOPES.RECORD_REGISTRATION_CORRECT])

  await expect(
    client.event.actions.correction.approve(
      generator.event.actions.correction.approve(
        'registered-event-test-id-12345',
        'request-test-id-12345'
      )
    )
  ).rejects.not.toMatchObject(new TRPCError({ code: 'FORBIDDEN' }))
})

test(`${ActionType.REJECT_CORRECTION} prevents forbidden access if missing required scope`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user, [])

  await expect(
    client.event.actions.correction.reject(
      generator.event.actions.correction.reject(
        'registered-event-test-id-12345',
        'request-test-id-12345'
      )
    )
  ).rejects.toMatchObject(new TRPCError({ code: 'FORBIDDEN' }))
})

test(`${ActionType.REJECT_CORRECTION} allows access if required scope is present`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user, [SCOPES.RECORD_REGISTRATION_CORRECT])

  await expect(
    client.event.actions.correction.reject(
      generator.event.actions.correction.reject(
        'registered-event-test-id-12345',
        'request-test-id-12345'
      )
    )
  ).rejects.not.toMatchObject(new TRPCError({ code: 'FORBIDDEN' }))
})

test('a correction request can be added to a created event', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const originalEvent = await client.event.create(generator.event.create())

  const declareInput = generator.event.actions.declare(originalEvent.id)

  await client.event.actions.declare.request(declareInput)

  const createAction = originalEvent.actions.filter(
    (action) => action.type === ActionType.CREATE
  )

  const assignmentInput = generator.event.actions.assign(originalEvent.id, {
    assignedTo: createAction[0].createdBy
  })

  await client.event.actions.assignment.assign(assignmentInput)

  const registeredEvent = await client.event.actions.register.request(
    generator.event.actions.register(originalEvent.id)
  )

  await client.event.actions.assignment.assign({
    ...assignmentInput,
    transactionId: getUUID()
  })
  const withCorrectionRequest = await client.event.actions.correction.request(
    generator.event.actions.correction.request(registeredEvent.id)
  )

  expect(withCorrectionRequest.actions.slice(-2)).toEqual([
    expect.objectContaining({
      type: ActionType.REQUEST_CORRECTION
    }),
    expect.objectContaining({ type: ActionType.UNASSIGN })
  ])
})

test(`${ActionType.REQUEST_CORRECTION} validation error message contains all the offending fields`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  const data = generator.event.actions.correction.request(event.id, {
    declaration: {
      'applicant.dob': '02-02',
      'applicant.dobUnknown': false,
      'recommender.none': true
    }
  })

  await expect(
    client.event.actions.correction.request(data)
  ).rejects.matchSnapshot()
})

test(`${ActionType.REQUEST_CORRECTION} when mandatory field is invalid, conditional hidden fields are still skipped`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  const data = generator.event.actions.correction.request(event.id, {
    declaration: {
      'applicant.dob': '02-1-2024',
      'applicant.dobUnknown': false,
      'applicant.name': {
        firstname: 'John',
        surname: 'Doe'
      },
      'recommender.none': true,
      'applicant.address': {
        country: 'FAR',
        addressType: AddressType.DOMESTIC,
        province: 'a45b982a-5c7b-4bd9-8fd8-a42d0994054c',
        district: '5ef450bc-712d-48ad-93f3-8da0fa453baa',
        urbanOrRural: 'RURAL' as const,
        village: 'Small village'
      }
    }
  })

  await expect(
    client.event.actions.correction.request(data)
  ).rejects.matchSnapshot()
})

test(`${ActionType.REQUEST_CORRECTION} Skips required field validation when they are conditionally hidden`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  const form = {
    'applicant.dob': '2024-02-01',
    'applicant.dobUnknown': false,
    'applicant.name': {
      firstname: 'John',
      surname: 'Doe'
    },
    'recommender.none': true,
    'applicant.address': {
      country: 'FAR',
      addressType: AddressType.DOMESTIC,
      province: 'a45b982a-5c7b-4bd9-8fd8-a42d0994054c',
      district: '5ef450bc-712d-48ad-93f3-8da0fa453baa',
      urbanOrRural: 'RURAL' as const,
      village: 'Small village'
    }
  }

  const data = generator.event.actions.correction.request(event.id, {
    declaration: form
  })

  const response = await client.event.actions.correction.request(data)
  const activeActions = getAcceptedActions(response)

  const savedAction = activeActions.find(
    (action) => action.type === ActionType.REQUEST_CORRECTION
  )
  expect(savedAction?.declaration).toEqual(form)
})

test(`${ActionType.REQUEST_CORRECTION} Prevents adding birth date in future`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  const form = {
    'applicant.dob': '2040-02-01',
    'applicant.dobUnknown': false,
    'applicant.name': {
      firstname: 'John',
      surname: 'Doe'
    },
    'recommender.none': true,
    'applicant.address': {
      country: 'FAR',
      addressType: AddressType.DOMESTIC,
      province: 'a45b982a-5c7b-4bd9-8fd8-a42d0994054c',
      district: '5ef450bc-712d-48ad-93f3-8da0fa453baa',
      urbanOrRural: 'RURAL' as const,
      village: 'Small village'
    }
  }

  const payload = generator.event.actions.correction.request(event.id, {
    declaration: form
  })

  await expect(
    client.event.actions.correction.request(payload)
  ).rejects.matchSnapshot()
})

test('a correction request can be added to a created event', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const originalEvent = await client.event.create(generator.event.create())

  await client.event.actions.declare.request(
    generator.event.actions.declare(originalEvent.id)
  )

  const createAction = originalEvent.actions.filter(
    (action) => action.type === ActionType.CREATE
  )

  const assignmentInput = generator.event.actions.assign(originalEvent.id, {
    assignedTo: createAction[0].createdBy
  })

  await client.event.actions.assignment.assign(assignmentInput)

  const registeredEvent = await client.event.actions.register.request(
    generator.event.actions.register(originalEvent.id)
  )

  await client.event.actions.assignment.assign({
    ...assignmentInput,
    transactionId: getUUID()
  })
  const withCorrectionRequest = await client.event.actions.correction.request(
    generator.event.actions.correction.request(registeredEvent.id)
  )

  expect(withCorrectionRequest.actions.slice(-2)).toEqual([
    expect.objectContaining({
      type: ActionType.REQUEST_CORRECTION
    }),
    expect.objectContaining({ type: ActionType.UNASSIGN })
  ])
})

describe('when a correction request exists', () => {
  let withCorrectionRequest: EventDocument
  let client: ReturnType<typeof createTestClient>

  const rng = createPrng(94842)

  beforeEach(async () => {
    const { user, generator } = await setupTestCase()
    client = createTestClient(user)

    const originalEvent = await client.event.create(generator.event.create())

    const declareInput = generator.event.actions.declare(originalEvent.id)

    await client.event.actions.declare.request(declareInput)

    const createAction = originalEvent.actions.filter(
      (action) => action.type === ActionType.CREATE
    )

    const assignmentInput = generator.event.actions.assign(originalEvent.id, {
      assignedTo: createAction[0].createdBy
    })

    await client.event.actions.assignment.assign(assignmentInput)

    const registeredEvent = await client.event.actions.register.request(
      generator.event.actions.register(originalEvent.id)
    )

    await client.event.actions.assignment.assign({
      ...assignmentInput,
      transactionId: getUUID()
    })
    withCorrectionRequest = await client.event.actions.correction.request(
      generator.event.actions.correction.request(registeredEvent.id, {
        declaration: {
          ...generateActionDeclarationInput(
            tennisClubMembershipEvent,
            ActionType.DECLARE,
            rng
          ),
          'applicant.name': {
            firstname: 'Johnny',
            surname: 'Doe'
          }
        }
      })
    )
    await client.event.actions.assignment.assign({
      ...assignmentInput,
      transactionId: getUUID()
    })
  })

  test('a correction request can be approved with correct request id', async () => {
    const { generator } = await setupTestCase()

    const requestId =
      withCorrectionRequest.actions[withCorrectionRequest.actions.length - 1].id

    const withApprovedCorrectionRequest =
      await client.event.actions.correction.approve(
        generator.event.actions.correction.approve(
          withCorrectionRequest.id,
          requestId
        )
      )
    expect(withApprovedCorrectionRequest.actions.slice(-2)).toEqual([
      expect.objectContaining({
        type: ActionType.APPROVE_CORRECTION,
        requestId
      }),
      expect.objectContaining({ type: ActionType.UNASSIGN })
    ])
  })

  test('approving a request fails if request id is incorrect', async () => {
    const { generator } = await setupTestCase()

    const incorrectRequestId = getUUID()

    const request = client.event.actions.correction.approve(
      generator.event.actions.correction.approve(
        withCorrectionRequest.id,
        incorrectRequestId
      )
    )
    await expect(request).rejects.toThrow()
  })
})

test(`${ActionType.APPROVE_CORRECTION} is idempotent`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const originalEvent = await client.event.create(generator.event.create())

  const declareInput = generator.event.actions.declare(originalEvent.id)

  await client.event.actions.declare.request(declareInput)

  const createAction = originalEvent.actions.filter(
    (action) => action.type === ActionType.CREATE
  )

  const assignmentInput = generator.event.actions.assign(originalEvent.id, {
    assignedTo: createAction[0].createdBy
  })

  await client.event.actions.assignment.assign(assignmentInput)

  const registeredEvent = await client.event.actions.register.request(
    generator.event.actions.register(originalEvent.id)
  )

  await client.event.actions.assignment.assign({
    ...assignmentInput,
    transactionId: getUUID()
  })

  const withCorrectionRequest = await client.event.actions.correction.request(
    generator.event.actions.correction.request(registeredEvent.id, {})
  )

  await client.event.actions.assignment.assign({
    ...assignmentInput,
    transactionId: getUUID()
  })

  const actionId = withCorrectionRequest.actions.at(-1)?.id

  if (!actionId) {
    throw new Error('Request ID is undefined')
  }

  const approveCorrectionPayload = generator.event.actions.correction.approve(
    withCorrectionRequest.id,
    actionId,
    { keepAssignment: true }
  )

  const firstResponse = await client.event.actions.correction.approve(
    approveCorrectionPayload
  )
  const secondResponse = await client.event.actions.correction.approve(
    approveCorrectionPayload
  )

  expect(firstResponse).toEqual(secondResponse)
})

test(`${ActionType.REJECT_CORRECTION} is idempotent`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const originalEvent = await client.event.create(generator.event.create())

  const declareInput = generator.event.actions.declare(originalEvent.id)

  await client.event.actions.declare.request(declareInput)

  const createAction = originalEvent.actions.filter(
    (action) => action.type === ActionType.CREATE
  )

  const assignmentInput = generator.event.actions.assign(originalEvent.id, {
    assignedTo: createAction[0].createdBy
  })

  await client.event.actions.assignment.assign(assignmentInput)

  const registeredEvent = await client.event.actions.register.request(
    generator.event.actions.register(originalEvent.id)
  )

  await client.event.actions.assignment.assign({
    ...assignmentInput,
    transactionId: getUUID()
  })

  const withCorrectionRequest = await client.event.actions.correction.request(
    generator.event.actions.correction.request(registeredEvent.id)
  )

  await client.event.actions.assignment.assign({
    ...assignmentInput,
    transactionId: getUUID()
  })

  const actionId = withCorrectionRequest.actions.at(-1)?.id

  if (!actionId) {
    throw new Error('Request ID is undefined')
  }

  const rejectCorrectionPayload = generator.event.actions.correction.reject(
    withCorrectionRequest.id,
    actionId,
    { keepAssignment: true }
  )

  const firstResponse = await client.event.actions.correction.reject(
    rejectCorrectionPayload
  )
  const secondResponse = await client.event.actions.correction.reject(
    rejectCorrectionPayload
  )

  expect(firstResponse).toEqual(secondResponse)
})

test('a correction request is not allowed if the event is already waiting for correction', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  await client.event.actions.correction.request(
    generator.event.actions.correction.request(event.id)
  )

  const createAction = event.actions.filter(
    (action) => action.type === ActionType.CREATE
  )

  const assignmentInput = generator.event.actions.assign(event.id, {
    assignedTo: createAction[0].createdBy
  })

  await client.event.actions.assignment.assign(assignmentInput)

  await expect(
    client.event.actions.correction.request(
      generator.event.actions.correction.request(event.id)
    )
  ).rejects.toThrow(
    new TRPCError({
      code: 'CONFLICT',
      message: 'Event is waiting for correction'
    })
  )
})
