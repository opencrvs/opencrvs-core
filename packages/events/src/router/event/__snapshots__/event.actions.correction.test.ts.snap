// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`REQUEST_CORRECTION Prevents adding birth date in future 1`] = `[TRPCError: [{"message":"Please enter a valid date","id":"applicant.dob","value":"2040-02-01"}]]`;

exports[`REQUEST_CORRECTION prevents correcting a field which is configured as uncorrectable: true 1`] = `[TRPCError: [{"message":"Correction not allowed for field","id":"applicant.email","value":"john.doe@example.com"}]]`;

exports[`REQUEST_CORRECTION validation error message contains all the offending fields 1`] = `[TRPCError: [{"message":"Invalid date field","id":"applicant.dob","value":"02-02"},{"message":"Please enter a valid date","id":"applicant.dob","value":"02-02"}]]`;

exports[`REQUEST_CORRECTION when mandatory field is invalid, conditional hidden fields are still skipped 1`] = `[TRPCError: [{"message":"Invalid date field","id":"applicant.dob","value":"02-1-2024"},{"message":"Please enter a valid date","id":"applicant.dob","value":"02-1-2024"}]]`;
